pyasdf/fits_embed.py

```diff
@@ -0,0 +1,156 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+# -*- coding: utf-8 -*-
+
+from __future__ import absolute_import, division, unicode_literals, print_function
+
+"""
+Utilities for embedded ADSF files in FITS.
+"""
+import io
+import re
+
+import numpy as np
+
+from astropy.extern import six
+from astropy.io import fits
+
+from . import asdf
+from . import block
+from . import util
+
+
+ASDF_EXTENSION_NAME = 'ASDF'
+
+FITS_SOURCE_PREFIX = 'fits:'
+
+
+class _FitsBlock(object):
+    def __init__(self, hdu):
+        self._hdu = hdu
+
+    def __repr__(self):
+        return '<FitsBlock {0},{1}>'.format(self._hdu.name, self._hdu.ver)
+
+    def __len__(self):
+        return self._hdu.data.nbytes
+
+    @property
+    def data(self):
+        return self._hdu.data
+
+    @property
+    def array_storage(self):
+        return 'fits'
+
+
+class _EmbeddedBlockManager(block.BlockManager):
+    def __init__(self, hdulist, asdffile):
+        self._hdulist = hdulist
+
+        super(_EmbeddedBlockManager, self).__init__(asdffile)
+
+    def get_block(self, source):
+        if (isinstance(source, six.string_types) and
+            source.startswith(FITS_SOURCE_PREFIX)):
+            parts = re.match(
+                '((?P<name>[A-Z0-9]+),)?(?P<ver>[0-9]+)',
+                source[len(FITS_SOURCE_PREFIX):])
+            if parts is not None:
+                ver = int(parts.group('ver'))
+                if parts.group('name'):
+                    pair = (parts.group('name'), ver)
+                else:
+                    pair = ver
+                return _FitsBlock(self._hdulist[pair])
+            else:
+                raise ValueError("Can not parse source '{0}'".format(source))
+
+        return super(_EmbeddedBlockManager, self).get_block(source)
+
+    def get_source(self, block):
+        if isinstance(block, _FitsBlock):
+            for i, hdu in enumerate(self._hdulist):
+                if hdu is block._hdu:
+                    if hdu.name == '':
+                        return '{0}{1}'.format(
+                            FITS_SOURCE_PREFIX, i)
+                    else:
+                        return '{0}{1},{2}'.format(
+                            FITS_SOURCE_PREFIX, hdu.name, hdu.ver)
+            raise ValueError("FITS block seems to have been removed")
+
+        return super(_EmbeddedBlockManager, self).get_source(block)
+
+    def find_or_create_block_for_array(self, arr, ctx):
+        from .tags.core import ndarray
+
+        if not isinstance(arr, ndarray.NDArrayType):
+            base = util.get_array_base(arr)
+            for hdu in self._hdulist:
+                if base is hdu.data:
+                    return _FitsBlock(hdu)
+
+        return super(
+            _EmbeddedBlockManager, self).find_or_create_block_for_array(arr, ctx)
+
+
+class AsdfInFits(asdf.AsdfFile):
```

Comment:
You can grab the YAML content by pulling out the data from the `ASDF` HDU.  But this class will be necessary in order to make the data linkages work.

That should be in an example, but it would basically be:

```
with fits.open("test.fits") as hdulist:
    with AsdfInFits.open(hdulist) as ff:
        ff.tree[...]
```